File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const n = maps.length;
    const m = maps[0].length;
    const dx = [-1, 0, 1, 0];
    const dy = [0, 1, 0, -1];
    
    const visited = Array.from(Array(n), () => new Array(m).fill(false));
    const dist = Array.from(Array(n), () => new Array(m).fill(0));
    
    let q = new Array(n*m);
    let head = 0, tail = 0;
    q[tail++] = [0, 0];
    
    visited[0][0] = true;
    dist[0][0] = 1;
    
    while(head < tail) {
        const [curX, curY] = q[head++];
        if (curX === n - 1 && curY === m - 1) return dist[curX][curY];
        for (let i = 0; i < 4; i ++) {
            const x = curX + dx[i];
            const y = curY + dy[i];
            if (x < 0 || x >= n || y < 0 || y >= m) continue;
            if(maps[x][y] !== 1 || visited[x][y]) continue;
            visited[x][y] = true;
            dist[x][y] = dist[curX][curY] + 1;
            q[tail++] = [x, y];
        }
    }
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n, m = len(maps), len(maps[0])
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    
    q = deque()
    q.append((0, 0))
    
    while q:
        x, y = q.popleft()
        
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            
            if not (0 <= nx < n and 0 <= ny < m) or maps[nx][ny] == 0:
                continue
            
            if maps[nx][ny] == 1:
                maps[nx][ny] = maps[x][y] + 1
                q.append((nx, ny))
    
    return maps[n - 1][m - 1] if maps[n - 1][m - 1] != 1 else -1